File: KLine-Chart-OC/KLine-Chart-OC/KLineChart/KLineChartView.m
//
//  KLineChartView.m
//  KLine-Chart-OC
//
//  Created by 何俊松 on 2020/3/10.
//  Copyright © 2020 hjs. All rights reserved.
//

#import "KLineChartView.h"
#import "ChartStyle.h"
#import "KLinePainterView.h"
#import "KLineInfoView.h"

@interface KLineChartView()
@property(nonatomic,strong) KLinePainterView *painterView;
@property(nonatomic,strong) KLineInfoView *infoView;

@property(nonatomic,assign) CGFloat maxScroll;
@property(nonatomic,assign) CGFloat minScroll;

@property(nonatomic,assign) CGFloat lastScrollX;


@property(nonatomic,assign) CGFloat dragbeginX;
@property(nonatomic,assign) BOOL isDrag;
@property(nonatomic,assign) CGFloat speedX;
@property(nonatomic,strong) CADisplayLink *displayLink;

@property(nonatomic,assign) BOOL isScale;
@property(nonatomic,assign) CGFloat lastscaleX;

@end


@implementation KLineChartView

- (void)setFrame:(CGRect)frame {
    [super setFrame:frame];
    self.painterView.frame = self.bounds;
    [self initIndicatirs];
}

- (void)setDatas:(NSArray<KLineModel *> *)datas {
    _datas = datas;
    [self initIndicatirs];
    self.painterView.datas = datas;
}

- (void)setIsLine:(BOOL)isLine {
    _isLine = isLine;
    self.painterView.isLine = isLine;
}

- (void)setIsLongPress:(BOOL)isLongPress {
    _isLongPress = isLongPress;
    self.painterView.isLongPress = isLongPress;
    if(!isLongPress) {
        [self.infoView removeFromSuperview];
    }
}

-(void)setScrollX:(CGFloat)scrollX {
    _scrollX = scrollX;
    self.painterView.scrollX = scrollX;
}

- (void)setScaleX:(CGFloat)scaleX {
    _scaleX = scaleX;
    [self initIndicatirs];
    self.painterView.scaleX = scaleX;
}

- (void)setMainState:(MainState)mainState {
    _mainState = mainState;
    self.painterView.mainState = mainState;
}

-(void)setSecondaryState:(SecondaryState)secondaryState {
    _secondaryState = secondaryState;
    self.painterView.secondaryState = secondaryState;
}

-(void)setLongPressX:(CGFloat)longPressX {
    _longPressX = longPressX;
    self.painterView.longPressX = longPressX;
}

-(void)setDirection:(KLineDirection)direction {
    _direction = direction;
    self.painterView.direction = direction;
}


- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        _scaleX = 1;
        _mainState = MainStateMA;
        _secondaryState = SecondaryStateWR;
        _scrollX = -self.frame.size.width / 5 + ChartStyle_candleWidth / 2;
        [self initIndicatirs];
        _painterView = [[KLinePainterView alloc] initWithFrame:self.bounds datas:_datas scrollX:_scrollX isLine:_isLine scaleX:_scaleX isLongPress:_isLongPress mainState:_mainState secondaryState:_secondaryState];
        [self addSubview:_painterView];
         __weak typeof(self) weakSelf = self;
        _painterView.showInfoBlock = ^(KLineModel * _Nonnull model, BOOL isLeft) {
            weakSelf.infoView.model = model;
            [weakSelf addSubview:weakSelf.infoView];
            CGFloat padding = 5;
            if(isLeft){
                weakSelf.infoView.frame = CGRectMake(padding, 30,  weakSelf.infoView.frame.size.width,  weakSelf.infoView.frame.size.height);
            } else {
                weakSelf.infoView.frame = CGRectMake(weakSelf.frame.size.width - weakSelf.infoView.frame.size.width - padding, 30,  weakSelf.infoView.frame.size.width,  weakSelf.infoView.frame.size.height);
            }
        };
        UIPanGestureRecognizer *panGesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(dragKlineEvent:)];
        UILongPressGestureRecognizer *longGresture = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(longPressKlineEvent:)];
        UIPinchGestureRecognizer *pinGesture = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(secalXEvent:)];
        [_painterView addGestureRecognizer:panGesture];
        [_painterView addGestureRecognizer:longGresture];
        [_painterView addGestureRecognizer:pinGesture];
    }
    return self;
}


-(void)initIndicatirs {
    CGFloat dataLength = ((CGFloat)_datas.count) * (ChartStyle_candleWidth * _scaleX + ChartStyle_canldeMargin) - ChartStyle_canldeMargin;
      _maxScroll = dataLength - self.frame.size.width;
//    if(dataLength > self.frame.size.width) {
//        _maxScroll = dataLength - self.frame.size.width;
//    } else {
//        _maxScroll =  -(self.frame.size.width - dataLength);
//    }
    CGFloat dataScroll = self.frame.size.width - dataLength;
    CGFloat normalminScroll = -self.frame.size.width/5.0 + (ChartStyle_candleWidth * _scaleX) / (CGFloat)2.0;
    self.minScroll = MIN(normalminScroll,-dataScroll);
    self.scrollX = [self clamp:_scrollX min:_minScroll max:_maxScroll];
    self.lastScrollX = self.scrollX;
    
}

-(void)dragKlineEvent:(UIPanGestureRecognizer *)gesture{
    if(_displayLink) {
        [_displayLink invalidate];
        _displayLink = nil;
    }
    switch (gesture.state) {
        case UIGestureRecognizerStateBegan:
        {
            CGPoint point = [gesture locationInView:self.painterView];
            _dragbeginX = point.x;
            _isDrag = true;
        } break;
        case UIGestureRecognizerStateChanged:
        {
            CGPoint point = [gesture locationInView:self.painterView];
            CGFloat dragX = point.x - _dragbeginX;
            self.scrollX = [self clamp:_lastScrollX + dragX min:_minScroll max:_maxScroll];
        } break;
        case UIGestureRecognizerStateEnded:
        {
            CGPoint speed = [gesture velocityInView:self.painterView];
            self.speedX = speed.x;
            _isDrag = false;
            self.lastScrollX = self.scrollX;
            if(speed.x != 0) {
                _displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(refreshEvent:)];
                [_displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSRunLoopCommonModes];
            }
        }break;
        default:
            break;
    }
    
}

-(void)longPressKlineEvent:(UILongPressGestureRecognizer *)gesture {
    switch (gesture.state) {
        case UIGestureRecognizerStateBegan:
        {
            CGPoint point = [gesture locationInView:self.painterView];
            self.longPressX = point.x;
            self.isLongPress = YES;
        } break;
        case UIGestureRecognizerStateChanged:
        {
            CGPoint point = [gesture locationInView:self.painterView];
            self.longPressX = point.x;
            self.isLongPress = YES;
        } break;
        case UIGestureRecognizerStateEnded:
            self.isLongPress = NO;
        default:
            break;
    }
}
-(void)secalXEvent:(UIPinchGestureRecognizer *)gesture {
    switch (gesture.state) {
        case UIGestureRecognizerStateBegan:
            _isScale = true;
            break;
        case UIGestureRecognizerStateChanged:
        {
            _isScale = true;
            self.scaleX = [self clamp:self.lastscaleX * gesture.scale min:0.5 max:2];
        }
        case UIGestureRecognizerStateEnded:
        {
            _isScale = false;
            self.lastscaleX = _scaleX;
        }
        default:
            break;
    }
}

-(void)refreshEvent:(CADisplayLink *)displaylink {
    CGFloat space = 100;
    if(self.speedX < 0) {
        self.speedX = MIN(self.speedX + space,0);
        self.scrollX = [self clamp:self.scrollX - 5 min:_minScroll max:_maxScroll];
        self.lastscaleX = self.scrollX;
    } else if (self.speedX > 0) {
        self.speedX = MAX(self.speedX - space,0);
        self.scrollX = [self clamp:self.scrollX + 5 min:_minScroll max:_maxScroll];
        self.lastScrollX = self.scrollX;
    } else {
        [_displayLink invalidate];
        _displayLink = nil;
    }
}


-(CGFloat)clamp:(CGFloat)value min:(CGFloat)min max:(CGFloat)max {
    if (value < min) {
        return min;
    } else if (value > max) {
        return max;
    } else {
        return value;
    }
}

-(KLineInfoView *)infoView {
    if(_infoView == nil) {
        _infoView = [KLineInfoView lineInfoView];
    }
    return _infoView;
}


@end

File: KLine-Chart/KLine-Chart/KLineChart/KLineChartView.swift
//
//  KLineChartView.swift
//  KLine-Chart
//
//  Created by 何俊松 on 2020/3/1.
//  Copyright © 2020 hjs. All rights reserved.
//

import UIKit

public class KLineChartView: UIView {
    var painterView: KLinePainterView!
    var isLine = false {
        didSet {
            painterView.isLine = isLine
        }
    }

    var isScale = false
    var isDrag = false
    var isLongPress = false {
        didSet {
            painterView.isLongPress = isLongPress
            if !isLongPress {
                infoView.removeFromSuperview()
            }
        }
    }

    var scrollX: CGFloat = 0.0 {
        didSet {
            painterView.scrollX = scrollX
        }
    }

    var maxScroll: CGFloat = 0.0
    var minScroll: CGFloat = 0.0
    var scaleX: CGFloat = 1.0 {
        didSet {
            painterView?.scaleX = scaleX
        }
    }

    var selectX: CGFloat = 0.0
    var datas: [KLineModel] = [] {
        didSet {
            initIndicators()
            painterView.datas = datas
        }
    }

    var mainState: MainState = .none {
        didSet {
            painterView.mainState = mainState
        }
    }

    var secondaryState: SecondaryState = .none {
        didSet {
            painterView.secondaryState = secondaryState
        }
    }

    // 拖动相关参数
    var lastScrollX: CGFloat = 0.0
    var dragbeginX: CGFloat = 0

    /// 上一次缩放的比例
    var lastscaleX: CGFloat = 1
    /// 缩放的中心点
    var scaleStartX: CGFloat = 0
    /// 缩放中心点对应的index
    var scaleIndex: Int = 0

    // k线初始位置
    var minTrailing: CGFloat = 0
    var maxTrailing: CGFloat = 0

    var longPressX: CGFloat = 0 {
        didSet {
            painterView.longPressX = longPressX
        }
    }

    var direction: KLineDirection = .vertical {
        didSet {
            painterView.direction = direction
        }
    }

    var speedX: CGFloat = 0

    var displayLink: CADisplayLink?

    lazy var infoView: KLineInfoView = {
        let view = KLineInfoView.lineInfoView()
        return view
    }()

    public override var frame: CGRect {
        didSet {
            self.painterView?.frame = self.bounds
            initIndicators()
        }
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        painterView = KLinePainterView(
            frame: bounds,
            datas: datas,
            scrollX: scrollX,
            isLine: isLine,
            scaleX: scaleX,
            isLongPress: isLongPress,
            mainState: mainState,
            secondaryState: secondaryState
        )
        initTrailing()
        scrollX = -self.frame.width / 5 + ChartStyle.candleWidth / 2
        initIndicators()
        addSubview(painterView)
        painterView.showInfoBlock = {
            [weak self] point, isleft in
            guard let strongSelf = self else { return }
            strongSelf.infoView.model = point
            strongSelf.addSubview(strongSelf.infoView)
            let padding: CGFloat = 5
            if isleft {
                strongSelf.infoView.frame = CGRect(x: padding, y: 30, width: strongSelf.infoView.frame.width, height: strongSelf.infoView.frame.height)
            } else {
                strongSelf.infoView.frame = CGRect(x: strongSelf.frame.width - strongSelf.infoView.frame.width - padding, y: 30, width: strongSelf.infoView.frame.width, height: strongSelf.infoView.frame.height)
            }
        }
        let panGesture = UIPanGestureRecognizer(
            target: self,
            action: #selector(dragKlineEvent(gesture:))
        )
        painterView.addGestureRecognizer(panGesture)
        let longPressGreture = UILongPressGestureRecognizer(
            target: self,
            action: #selector(longPressKlineEvent(gesture:))
        )
        painterView.addGestureRecognizer(longPressGreture)
        let pinGesture = UIPinchGestureRecognizer(
            target: self,
            action: #selector(secalXEvent(gesture:))
        )
        painterView.addGestureRecognizer(pinGesture)
    }

    required init?(coder _: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    func initTrailing() {
        minTrailing = frame.width * 0.2
        maxTrailing = frame.width * 0.2
    }

    func initScrollWidth() {
        let dataLength: CGFloat = CGFloat(datas.count) * (ChartStyle.candleWidth * scaleX + ChartStyle.canldeMargin) - ChartStyle.canldeMargin
        if dataLength > frame.width {
            maxScroll = dataLength - frame.width
        } else {
            maxScroll = -(frame.width - dataLength)
        }
        let dataScroll = frame.width - dataLength
        let normalminScroll = -maxTrailing + (ChartStyle.candleWidth * scaleX) / 2
        minScroll = min(normalminScroll, -dataScroll)
    }

    func secalXIndicators(trailing: CGFloat) {
        initScrollWidth()
        let selectIndexOffset = CGFloat(scaleIndex) * (ChartStyle.candleWidth * scaleX + ChartStyle.canldeMargin) - ChartStyle.canldeMargin
        scrollX = clamp(
            value: selectIndexOffset - trailing,
            min: minScroll,
            max: maxScroll
        )
        print(scrollX)
    }

    func initIndicators() {
        if datas.count == 0 { return }
        initScrollWidth()
        scrollX = clamp(value: scrollX, min: minScroll, max: maxScroll)
        lastScrollX = scrollX
    }

    // 拖动k线处理事件
    @objc func dragKlineEvent(gesture: UIPanGestureRecognizer) {
        switch gesture.state {
        case .began:
            let point = gesture.location(in: painterView)
            dragbeginX = point.x
            print("dragKlineEvent began")
            isDrag = true
        case .changed:
            let point = gesture.location(in: painterView)
            let dragX = point.x - dragbeginX
            scrollX = clamp(
                value: lastScrollX + dragX,
                min: minScroll,
                max: maxScroll
            )
            print(scrollX)
        case .ended:
            let speed = gesture.velocity(in: gesture.view)
            speedX = speed.x
            print("speed=\(speed)")
            isDrag = false
            lastScrollX = scrollX
            if speed.x != 0 {
                displayLink = CADisplayLink(
                    target: self,
                    selector: #selector(refreshEvent)
                )
                displayLink?.add(
                    to: RunLoop.current,
                    forMode: RunLoop.Mode.common
                )
            }
        default:
            print("拖动k线出现\(gesture.state)事件")
        }
    }

    // 长按手势处理
    @objc func longPressKlineEvent(gesture: UILongPressGestureRecognizer) {
        print("longPressKlineEvent")
        switch gesture.state {
        case .began:
            let point = gesture.location(in: painterView)
            longPressX = point.x
            isLongPress = true
        case .changed:
            let point = gesture.location(in: painterView)
            longPressX = point.x
            isLongPress = true
        case .ended:
            isLongPress = false
        default:
            print("长按k线出现\(gesture.state)事件")
        }
    }

    @objc func secalXEvent(gesture: UIPinchGestureRecognizer) {
        switch gesture.state {
        case .began:
            isScale = true
            let point = gesture.location(in: self)
            scaleStartX = point.x
            guard let index = calculateIndex(selectX: scaleStartX) else { return }
            scaleIndex = index
        case .changed:
            isScale = true
            let newscaleX = clamp(value: lastscaleX * gesture.scale, min: 0.1, max: 5)
            if newscaleX == scaleX { return }
            scaleX = newscaleX
            let rightWidth = frame.width - scaleStartX
            secalXIndicators(trailing: rightWidth)

        case .ended, .cancelled:
            isScale = false
            lastscaleX = scaleX
            lastScrollX = scrollX
        default:
            print("长按k线出现\(gesture.state)事件")
        }
    }

    @objc func refreshEvent(_: CADisplayLink) {
        let space: CGFloat = 100
        if speedX < 0 {
            speedX = min(speedX + space, 0)
            scrollX = clamp(
                value: scrollX - 5,
                min: minScroll,
                max: maxScroll
            )
            lastScrollX = scrollX
        } else if speedX > 0 {
            speedX = max(speedX - space, 0)
            scrollX = clamp(
                value: scrollX + 5,
                min: minScroll,
                max: maxScroll
            )
            lastScrollX = scrollX
        } else {
            displayLink?.invalidate()
            displayLink = nil
        }
    }

    private func calculateIndex(selectX _: CGFloat) -> Int? {
        return painterView?.calculateIndex(selectX: scaleStartX)
    }
}
